homeassistant/components/automation/template.py

```diff
@@ -24,13 +24,14 @@
     @callback
     def template_listener(entity_id, from_s, to_s):
         """Listen for state changes and calls action."""
+        context = to_s.context if to_s else None
```

Comment:
I agree that this is a bug that is being fixed, however:
 - probably better in a new PR 
 - better to listen to the raw state changed events, as those will always contain the right context, even when we delete

Reply:
This is really just there to get this through. The issue really is the first fire of the listener doesn't really have a `entity_id` responsible. I don't actually think it should fire at all, the caller should check it first and go from there. The way it was coded was pretty dodgy anyhow, it was basically firing on the first change of *any* state and using that state as the trigger.

Alternately there shouldn't be any arguments to the listener, because really they don't make a lot of sense.  Say your template is `{{ states.binary_sensor.sensor1 and states.binary_sensor.sensor2 }}` then it's pretty irrelevant which of those states actually triggered it. If sensor1 is true, then false, then sensor2 is true, then sensor1, then the 'triggering entity` will be sensor1, but it's all a bit irrelevant.

Reply:
I think that the arguments are relevant in that it tells the caller exactly which state change caused the callback to be triggered.

Reply:
Yep, I don't disagree exactly :grin: What about the initial fire however? 

The way this listener is defined, it fires whenever the template's result goes from false to true. If it's initially true but we coerce it to be false, then it's going to fire at some variable time in the future when *some entity* changes state at some stage, and then that random entity is going to be the entity "responsible" for the event firing. This doesn't make a lot of sense to me really.

My preferred option is that the initial fire doesn't happen, and the caller deals with the situation internally.